### ISSUE TITLE
Buggy imap_search/encoding on Microsoft Exchange/Office365

### ISSUE THREAD

**Issue by dico** [barbushin/php-imap#128]
Hi

I have found some kind of issue. Tested on Office365 and local Microsoft Exchange server.

If serverEncoding is sent as an empty string when creating the object, the imap_search would fail and return empty array. The string would look like this: _imap_search(Resource id #72, FROM "mail@address.com", SE_UID, );_ - with a comma and no encoding at the end.

In some weird way, the imap_search criteria "ALL" seems to work in this case, but no other criteria-parameters.

Also if I don't send any encoding parameter at all, UTF-8 would be set as default. I don't know why or what kind of encoding is used by Microsoft - but UTF-8 has the same problem, criteria "ALL" work, but no other criterias.

If I just remove ", $this->serverEncoding" on line 174, all imap_search criterias works.

Suggested solution: Don't set any server encoding as default and/or check if server-encoding parameter is empty.

**Comment by stevebauman** [comment]
I have the same issue, do you know what type of encoding to use with Exchange and Office 365?

**Comment by DonCallisto** [comment]
Same issue here. @barbushin is this onto a `todo` list, do you recognize the issue, do you need help? Please let us know :)

**Comment by barbushin** [comment]
Sorry guys, I'm just too busy in last few months. It will be great if somebody can find solution and send me Pull request. Cheers!

**Comment by DonCallisto** [comment]
If someone is interested in this kind of situation - I will tackle it, if I ever find time, later - I've tried to understood what's goes on here as PHP itself returns a `false` either for "bad request" than for "no results" (a very poor choice in my humble opinion).

BTW I've played around with python and what I got is

```
>>> import imaplib;
>>> M = imaplib.IMAP4_SSL('outlook.office365.com', 993)
>>> M.login('my@mail.com, 'myPassword')
('OK', ['LOGIN completed.'])
>>> M.select('INBOX')
('OK', ['18'])
>>> M.search('utf-8', '(SINCE "23-May-2016" BEFORE "24-May-2016")')
('NO', ['[BADCHARSET (US-ASCII)] The specified charset is not supported.'])
>>> M.search(None, '(SINCE "23-May-2016" BEFORE "24-May-2016")')
('OK', ['17 18'])
```

**Comment by a7lan** [comment]
Any news?  I have same problem too

**Comment by DonCallisto** [comment]
@a7lan Unfortunately, no.

**Comment by littleylv** [comment]
@barbushin You can take a look at [here](https://github.com/emersion/go-imap/issues/48). Maybe we can do what they said: "When searching fall back to US-ASCII if BADCHARSET is received."

**Comment by mrefaie** [comment]
@littleylv You are right the US-ASCII encoding is the solutions for that. Every one who faces this problem can try to intialize the mailbox with US-ASCII encoding.
`$mailbox = new PhpImap\Mailbox('{imap.gmail.com:993/imap/ssl}INBOX', 'some@gmail.com', '*********', __DIR__, 'US-ASCII');`

**Comment by mrefaie** [comment]
With that solution a search in the mailbox like that `$mailbox->searchMailbox(('UNFLAGGED FROM "Sender Name <sender@gmail.com>"'));` will not work. You have to use only the username `$mailbox->searchMailbox(('UNFLAGGED FROM "Sender Name"'));`

**Comment by commanddotcom** [comment]
US-ASCII leads to the issue with gmail accounts:
Error: 

>> Mime string encoding conversion failed

**Comment by yusufalakavuk** [comment]
I had the same issue while converting iso-8859-1 to UTF-8 and added following lines 
		if(!$convertedString ) {
			$convertedString = @utf8_encode($string);
		}
just before 
		if(!$convertedString) {
			throw new Exception('Mime string encoding conversion failed');
		}

and it works fine for me now.

**Comment by NormanEdance** [comment]
See fixes: [Problem with serverEncoding #101](https://github.com/barbushin/php-imap/issues/101)
Btw: duplicate :dancing_men:

**Comment by Sebi94nbg** [comment]
This is an issue on Exchange servers and not related to this library. However, we're still checking, if we can implement a workaround to get it somehow working. See #306.
